cpp/src/strings/json/json_path.cu

```diff
@@ -0,0 +1,917 @@
+                                       char* out_buf,
+                                       bitmask_type* out_validity,
+                                       size_type num_rows)
+{
+  uint64_t const tid = threadIdx.x + (blockDim.x * blockIdx.x);
+
+  bool is_valid = false;
+  if (tid < num_rows) {
+    size_type const src_size = offsets[tid + 1] - offsets[tid];
+    size_type output_size    = 0;
+    if (src_size > 0) {
+      char* dst             = out_buf ? out_buf + output_offsets[tid] : nullptr;
+      size_t const dst_size = out_buf ? output_offsets[tid + 1] - output_offsets[tid] : 0;
+
+      parse_result result;
+      json_output out;
+      thrust::tie(result, out) =
+        get_json_object_single(chars + offsets[tid], src_size, commands, dst, dst_size);
+      output_size = out.output_len;
+      if (out.element_count > 0 && result == parse_result::SUCCESS) { is_valid = true; }
+    }
+
+    // filled in only during the precompute step
+    if (!out_buf) { output_offsets[tid] = static_cast<size_type>(output_size); }
+  }
+
+  // validity filled in only during the output step
+  if (out_validity) {
+    uint32_t mask = __ballot_sync(0xffffffff, is_valid);
+    // 0th lane of the warp writes the validity
+    if (!(tid % cudf::detail::warp_size) && tid < num_rows) {
+      out_validity[cudf::word_index(tid)] = mask;
+    }
+  }
+}
+
+/**
+ * @copydoc cudf::strings::detail::get_json_object
+ */
+std::unique_ptr<cudf::column> get_json_object(cudf::strings_column_view const& col,
+                                              cudf::string_scalar const& json_path,
+                                              rmm::cuda_stream_view stream,
+                                              rmm::mr::device_memory_resource* mr)
+{
+  // preprocess the json_path into a command buffer
+  std::tuple<rmm::device_uvector<path_operator>, int, bool> preprocess =
+    build_command_buffer(json_path, stream);
+  CUDF_EXPECTS(std::get<1>(preprocess) <= max_command_stack_depth,
+               "Encountered json_path string that is too complex");
+
+  // allocate output offsets buffer.
+  auto offsets = cudf::make_fixed_width_column(
+    data_type{type_id::INT32}, col.size() + 1, mask_state::UNALLOCATED, stream, mr);
+  cudf::mutable_column_view offsets_view(*offsets);
+
+  // if the query is empty, return a string column containing all nulls
+  if (std::get<2>(preprocess)) {
+    thrust::generate(rmm::exec_policy(stream),
+                     offsets_view.head<size_type>(),
+                     offsets_view.head<size_type>() + offsets_view.size(),
+                     [] __device__() { return 0; });
+    return cudf::make_strings_column(
+      col.size(),
+      std::move(offsets),
+      cudf::make_fixed_width_column(
+        data_type{type_id::INT8}, 0, mask_state::UNALLOCATED, stream, mr),
+      col.size(),
+      cudf::detail::create_null_mask(col.size(), mask_state::ALL_NULL, stream, mr),
+      stream,
+      mr);
```

Comment:
This seems like a lot of data behind an all-null validity mask. I don't think the children should be necessary here.
Accessing the data for a null-element is undefined so no code should be accessing the children.
How about this instead?
```
 return std::make_unique<column>(
      data_type{type_id::STRING},
      col.size(),
      rmm::device_buffer{0, stream, mr},  // no data
      cudf::detail::create_null_mask(col.size(), mask_state::ALL_NULL, stream, mr),
      col.size());

```

Reply:
So this actually ends up confusing the column equivalence comparison.   It creates this situation:
- A column with a size() > 0
- Everything is null
- No children.

The column equivalence checks this:
```
    if (check_exact_equality || lhs.size() > 0) {
      EXPECT_EQ(lhs.num_children(), rhs.num_children());
    }
```
This makes sense because I think we've only ever seen strings columns with no child in cases where size == 0.  I suppose 
the check above could be modified to something like:
```
    if (check_exact_equality || (lhs.size() > 0 && lhs.null_count() > 0)) {
      EXPECT_EQ(lhs.num_children(), rhs.num_children());
    }
```

Feels mildly spooky to change that during burndown though.  @jrhemstad   Thoughts?

Reply:
Ok. I think that check should be fixed. I know I have other places that return all-null strings column with no children.